packages/tokamak/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol

```diff
@@ -1,48 +1,73 @@
 // SPDX-License-Identifier: MIT
 pragma solidity 0.8.15;
 
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 import { Predeploys } from "src/libraries/Predeploys.sol";
 import { OptimismPortal } from "src/L1/OptimismPortal.sol";
 import { CrossDomainMessenger } from "src/universal/CrossDomainMessenger.sol";
 import { ISemver } from "src/universal/ISemver.sol";
 import { Constants } from "src/libraries/Constants.sol";
+import { SafeCall } from "src/libraries/SafeCall.sol";
+import { Hashing } from "src/libraries/Hashing.sol";
+import { Encoding } from "src/libraries/Encoding.sol";
+import { Constants } from "src/libraries/Constants.sol";
 
 /// @custom:proxied
 /// @title L1CrossDomainMessenger
 /// @notice The L1CrossDomainMessenger is a message passing interface between L1 and L2 responsible
 ///         for sending and receiving data on the L1 side. Users are encouraged to use this
 ///         interface instead of interacting with lower-level contracts directly.
 contract L1CrossDomainMessenger is CrossDomainMessenger, ISemver {
+    using SafeERC20 for IERC20;
+
     /// @notice Address of the OptimismPortal. The public getter for this
     ///         is legacy and will be removed in the future. Use `portal()` instead.
     /// @custom:network-specific
     /// @custom:legacy
     OptimismPortal public PORTAL;
 
+    /// @notice Address of TON (ERC-20 token)
+    address public l1TONAddress;
+
     /// @notice Semantic version.
     /// @custom:semver 1.7.1
     string public constant version = "1.7.1";
 
     /// @notice Constructs the L1CrossDomainMessenger contract.
     constructor() CrossDomainMessenger(Predeploys.L2_CROSS_DOMAIN_MESSENGER) {
-        initialize({ _portal: OptimismPortal(payable(0)) });
+        initialize({ _portal: OptimismPortal(payable(0)),  _l1TONAddress: address(0) });
     }
 
     /// @notice Initializes the contract.
     /// @param _portal Address of the OptimismPortal contract on this network.
-    function initialize(OptimismPortal _portal) public reinitializer(Constants.INITIALIZER) {
+    function initialize(OptimismPortal _portal, address _l1TONAddress) public reinitializer(Constants.INITIALIZER) {
         PORTAL = _portal;
+        l1TONAddress = _l1TONAddress;
         __CrossDomainMessenger_init();
+        if (address(PORTAL) != address(0) && address(l1TONAddress) != address(0)) {
+            IERC20(l1TONAddress).approve(address(PORTAL), 2**256 - 1);
+        }
```

Comment:
I think we don't need to approve for the PORTAL address maximum amount right here, we can move to the `_sendMessage` function and only approve each time this function is called.

I mean that
```solidity
function _sendMessage(address _to, uint64 _gasLimit, uint256 _value, bytes memory _data) internal override{
        require(msg.value == 0, "Deny depositing ETH");
        if (_value > 0) {
            IERC20(l1TONAddress).safeTransferFrom(msg.sender, address(this), _value);
            bool isApproveSuccess = IERC20(l1TONAddress).approve(address(PORTAL), _value)
            require(isApproveSuccess, "the approval for L1 TON to Portal failed");
        }
        PORTAL.depositTransaction(_to, _value, _gasLimit, false, _data);
    }
```

I am not sure that if we approve the maximum amount of TON for PORTAL, there are any potential security issues here.

Reply:
Okay, It's my fault,,, 
I didn't know much about approve and couldn't think deeply about it. Actually, I didn't know that if it use the approved amount, it authority will be lost.
I'm sorry @nguyenzung 
Can we change it back the way you implement initially.

Reply:
It is okay, I think the above solution is secure enough because L1StandardBridge and L1CDM does not hold any TON after finishing the transactions like depositERC20 or sendMessage and those contracts know other.

Reply:
But It seems to consider here as well. There are able to remain some TON token in L1CDM... 🙇🏼‍♂️. I'm sorry again. You were right. I considered only code simplification before... 😢🙇🏼‍♂️

Reply:
You are right. It is possible to have TON in L1CDM. 
I am also rollback to the previous solution.

Reply:
Thank you! But I think it seems good to do it after review of contracts team!  We may add approve(0),,,